src/uct/sm/cma/cma_ep.c

```diff
@@ -55,67 +55,50 @@ ucs_status_t uct_cma_ep_common_zcopy(uct_ep_h tl_ep,
-    size_t iov_slice_length;
-    size_t iov_slice_delivered;
-    size_t local_iov_it;
-    size_t length = 0;
+    size_t local_iov_cnt;
+    size_t length;
+    size_t cur_iov_cnt;
     struct iovec local_iov[UCT_SM_MAX_IOV];
     struct iovec remote_iov;
-    uct_cma_ep_t *ep = ucs_derived_of(tl_ep, uct_cma_ep_t);
-
-    do {
-        iov_it_length = 0;
-        local_iov_it = 0;
-        for (iov_it = 0; iov_it < ucs_min(UCT_SM_MAX_IOV, iovcnt); ++iov_it) {
-            iov_slice_delivered = 0;
 
-            /* Get length of the particular iov element */
-            iov_slice_length = uct_iov_get_length(iov + iov_it);
+    while (iovcnt - iov_idx) {
+        cur_iov_cnt   = ucs_min(iovcnt - iov_idx, UCT_SM_MAX_IOV);
+        local_iov_cnt = uct_iovec_fill_iov(local_iov, &iov[iov_idx],
+                                           cur_iov_cnt, &length);
+        ucs_assert(local_iov_cnt <= cur_iov_cnt);
 
-            /* Skip the iov element if no data */
-            if (!iov_slice_length) {
-                continue;
-            }
-            iov_it_length += iov_slice_length;
-
-            if (iov_it_length <= delivered) {
-                continue; /* Skip the iov element if transferred already */
-            } else {
-                /* Let's assume the iov element buffer can be delivered partially */
-                if ((iov_it_length - delivered) < iov_slice_length) {
-                    iov_slice_delivered = iov_slice_length - (iov_it_length - delivered);
-                }
-            }
+        iov_idx += cur_iov_cnt;
+        ucs_assert(iov_idx <= iovcnt);
 
-            local_iov[local_iov_it].iov_base = (void *)((char *)iov[iov_it].buffer +
-                                                        iov_slice_delivered);
-            local_iov[local_iov_it].iov_len  = iov_slice_length - iov_slice_delivered;
-            ++local_iov_it;
-        }
-        if (!delivered) {
-            length = iov_it_length; /* Keep total length of the iov buffers */
+        if (!length) {
+            continue; /* Nothing to deliver */
         }
 
-        if(!length) {
-            return UCS_OK; /* Nothing to deliver */
-        }
+        local_iov_idx = 0;
 
-        remote_iov.iov_base = (void *)(remote_addr + delivered);
-        remote_iov.iov_len  = length - delivered;
+        do {
+            remote_iov.iov_base = remote_ptr;
+            remote_iov.iov_len  = length;
 
-        ret = fn_p(ep->remote_pid, local_iov, local_iov_it, &remote_iov, 1, 0);
-        if (ret < 0) {
-            ucs_error("%s delivered %zu instead of %zu, error message %s",
-                      fn_name, delivered, length, strerror(errno));
-            return UCS_ERR_IO_ERROR;
-        }
+            ret = fn_p(ep->remote_pid, &local_iov[local_iov_idx],
+                       local_iov_cnt - local_iov_idx, &remote_iov, 1, 0);
+            if (ret < 0) {
+                ucs_error("%s failed to deliver %zu bytes, error message %s",
```

Comment:
probably will be a conflict with @brminich PR

Reply:
yes, it needs to be handled when @brminich  PR gets merged